man/man8/ulpatch.8
.\" Copyright (C) 2022-2026 Rong Tao (rtoax@foxmail.com)
.\"
.\" SPDX-License-Identifier: GPL-2.0-or-later
.\"
.TH ulpatch 8  "2022-10-01" "USER COMMANDS"
.SH NAME
ulpatch \- Userspace livepatch.

.SH SYNOPSIS
.B ulpatch
[\fI\,OPTION\/\fR]...

.SH DESCRIPTION
.\" Add any additional description here
.PP
ulpatch is the core tool of ULPatch, which is used to manage the patch of the target process, including patch loading, patch unloading, and multi-patch management.

This program is a basic command of ULPatch.

.SH REQUIREMENTS
bfd

.SH ARGUMENTS
.SS
\fB\-p\fR, \fB\-\-pid\fR [PID]
Specify a target process's PID.

.SS
\fB\-\-patch\fR [ULPATCH.ELF]
Specify a ulpatch.elf file to patch to target process.

.SS
\fB\-\-unpatch\fR
Unpatch a latest ulpatch from target process.

.SS
\fB\-\-map-pfx\fR
Display prefix of ulp in
.IR /proc/ [PID] /maps .

.SH COMMON ARGUMENTS
.SS
\fB\-\-log-level\fR[=\fI\,LEVEL\/\fR], \fB\-\-lv\fR[=\fI\,LEVEL\/\fR]
Specify a log level. The LEVEL could be number(see
.BR syslog (3)
) or string(debug,dbg,info,inf,notice,note,warning,warn,error,err,crit,alert,emerg).
For example, if you want to set log level to LOG_ERR, just:
.BR --lv=error .

.SS
\fB\-\-log-debug\fR
Set log level to DEBUG.

.SS
\fB\-\-log-error\fR
Set log level to ERROR.

.SS
\fB\-u\fR, \fB\-\-dry-run\fR
Don't actually run.

.SS
\fB\-v\fR[vvv...], \fB\-\-verbose\fR
Show verbose information.
You can set more \fBv\fR values to increase the verbose level and display more details.
For example, \fB-vvvvv\fR will set the verbose value to \fB4\fR, because there are 4 \fBv\fR's.

.SS
\fB\-h\fR, \fB\-\-help\fR
Show help information.

.SS
\fB\-V\fR, \fB\-\-version\fR
Show version information.

.SS
\fB\-F\fR, \fB\-\-force\fR
Force do something.

.SS
\fB\-\-info\fR
Print detailed information about features supported by the kernel and the ULPatch build. It is necessary to display this information when you are submitting a MR/PR.

.SH MACROS
.SS
.B __ULP_DEV
If you develope an ulpatch ulp file, you need to define
.B __ULP_DEV
and include
.B <ulpatch/meta.h>
header in your ulpatch source code. The
.BR ulpconfig (8)
.B --cflags
provides this macro.

.SH ULP DEMO
More to see the documentations in ulpatch source code tree.
.SS Program source
\&
.\" SRC BEGIN (tee.c)
.EX
#ifndef __ULP_DEV
#define __ULP_DEV 1
#endif
#include <stdio.h>
#include <ulpatch/asm.h>
#include <ulpatch/meta.h>

void ulpatch_print_hello_exit(void)
{
	printf("Patched.\\n");
	__ulp_builtin_exit(0xff);
}
ULPATCH_INFO(ulpatch_print_hello_exit, print_hello, "Rong Tao");
.EE
.\" SRC END

.SH ARCHITECTURES
List the CPU architectures that ULPatch supports and will support.
.TS
allbox;
lbx lb lb
l l l.
Architecture	Supported	Value
T{
.na
.nh
.BR x86_64
T}	YES	TODO
T{
.na
.nh
.BR aarch64
T}	YES	TODO
T{
.na
.nh
.BR loongarch64
T}	NO	-
T{
.na
.nh
.BR riscv64
T}	NO	-
.TE

.SH SUPPORT_TYPE

.TS
allbox;
lbx lb lb
l l l.
Execution	Patch-Type	Supported
T{
.na
.nh
.BR ET_EXEC
T}	Add-Variable	yes
T{
.na
.nh
.BR ET_EXEC
T}	Call-Library-Function	yes
T{
.na
.nh
.BR ET_DYN
T}	Add-Variable	not-yet
T{
.na
.nh
.BR ET_DYN
T}	Call-Library-Function	not-yet
.TE

.SH SELFTESTS
Use \fBulpatch_test\fR command to test.

.SH LINKS
.IP " 1." 4
aarch64 relocations
.RS 4
\%https://github.com/ARM-software/abi-aa/releases
\%https://github.com/ARM-software/abi-aa/blob/main/aaelf64/aaelf64.rst
\%https://docslib.org/doc/4448214/elf-for-the-arm-64-bit-architecture-aarch64

.SH OS
Linux

.SH STABILITY
Unstable - in development.

.SH AUTHOR
Written by Rong Tao

.SH SEE ALSO
.BR ulpinfo (8),
.BR ulftrace (8),
.BR ultask (8),
.BR ulpconfig (8)

.P
The descriptions of the following files in
.BR proc (5):
.IR /proc/ pid /maps ,
.IR /proc/ pid /map_files ,
and
.IR /proc/ pid /smaps .


CHANGELOG.md
Changelog
=========

Not release yet, don't use.


src/ulpatch.c
// SPDX-License-Identifier: GPL-2.0-or-later
/* Copyright (C) 2022-2026 Rong Tao */
#include <stdlib.h>
#include <getopt.h>
#include <stdio.h>
#include <stdbool.h>
#include <assert.h>
#include <errno.h>

#include "init.h"
#include "version.h"
#include "elf/elf-api.h"

#include "patch/patch.h"

#include "utils/log.h"
#include "utils/list.h"
#include "utils/compiler.h"
#include "task/task.h"
#include "cmds.h"

#include "args-common.c"


enum command {
	CMD_NONE,
	CMD_PATCH,
	CMD_UNPATCH,
} command_type = CMD_NONE;


static pid_t target_pid = -1;
static struct task_struct *target_task = NULL;
static char *patch_file = NULL;

enum {
	ARG_MIN = ARG_COMMON_MAX,
	ARG_PATCH,
	ARG_UNPATCH,
	ARG_MAP_PFX,
};

static const char *prog_name = "ulpatch";

int check_patch_file(const char *file);

static void ulpatch_args_reset(void)
{
	target_pid = -1;
	target_task = NULL;
	patch_file = NULL;
}

static int print_help(void)
{
	printf(
	"\n"
	" Usage: ulpatch [OPTION]... [FILE]...\n"
	"\n"
	" ulpatch is the core tool of ULPatch, which is used to manage the \n"
	" patch of the target process, including patch loading, patch unloading,\n"
	" and multi-patch management.\n"
	"\n"
	" Option argument:\n"
	"\n"
	"  -p, --pid [PID]     specify a process identifier(pid_t)\n"
	"\n"
	" Operate argument:\n"
	"\n"
	"  --patch  [PATCH]    patch an object file into target task, and patch\n"
	"                      the patch.\n"
	"  --unpatch           unpatch the latest ulpatch from target task.\n"
	"\n"
	" Display argument:\n"
	"\n"
	"  --map-pfx           display /proc/PID/maps prefix: '%s'.\n"
	"\n",
	PATCH_VMA_TEMP_PREFIX);
	print_usage_common(prog_name);
	cmd_exit_success();
	return 0;
}

static int parse_config(int argc, char *argv[])
{
	int ret;

	struct option options[] = {
		{ "pid",            required_argument, 0, 'p' },
		{ "patch",          required_argument, 0, ARG_PATCH },
		{ "unpatch",        no_argument,       0, ARG_UNPATCH },
		{ "map-pfx",        no_argument,       0, ARG_MAP_PFX },
		COMMON_OPTIONS
		{ NULL }
	};

	while (1) {
		int c;
		int option_index = 0;
		c = getopt_long(argc, argv, "p:"COMMON_GETOPT_OPTSTRING,
				options, &option_index);
		if (c < 0)
			break;

		switch (c) {
		case 'p':
			target_pid = atoi(optarg);
			break;
		case ARG_PATCH:
			command_type = CMD_PATCH;
			patch_file = strdup(optarg);
			break;
		case ARG_UNPATCH:
			command_type = CMD_UNPATCH;
			break;
		case ARG_MAP_PFX:
			printf("%s\n", PATCH_VMA_TEMP_PREFIX);
			cmd_exit_success();
			break;
		COMMON_GETOPT_CASES(prog_name, print_help, argv)
		default:
			print_help();
			cmd_exit(1);
			break;
		}
	}

	if (command_type == CMD_NONE) {
		fprintf(stderr, "Nothing to do, check -h, --help.\n");
		cmd_exit(1);
	}

	if (target_pid == -1) {
		fprintf(stderr, "Specify pid with -p, --pid.\n");
		cmd_exit(1);
	}

	if (!proc_pid_exist(target_pid)) {
		fprintf(stderr, "pid %d not exist.\n", target_pid);
		cmd_exit(1);
	}

	/* check patch file */
	if (command_type == CMD_PATCH) {
		ret = check_patch_file(patch_file);
		if (ret) {
			fprintf(stderr, "Check %s failed.\n", patch_file);
			cmd_exit(1);
		}
	}

	return 0;
}

int check_patch_file(const char *file)
{
	int err = 0;
	struct load_info info = {0};

	if (!file)
		return -EEXIST;

	if (file && !fexist(file)) {
		ulp_debug("%s is not exist.\n", file);
		return -EEXIST;
	}
	err = alloc_patch_file(patch_file, "temp.ulp", &info);
	if (err) {
		ulp_error("Parse %s failed.\n", patch_file);
		return err;
	}

	err = setup_load_info(&info);
	if (err) {
		ulp_debug("Load %s failed\n", file);
		err = -ENODATA;
		goto release;
	}

	if (strcmp(info.ulp_strtab.magic, SEC_ULPATCH_MAGIC)) {
		ulp_debug("%s is not ulpatch file.\n", file);
		err = -ENODATA;
	}

release:
	release_load_info(&info);
	return err;
}

static int command_patch(void)
{
	return init_patch(target_task, patch_file);
}

static int command_unpatch(void)
{
	return delete_patch(target_task);
}

int ulpatch(int argc, char *argv[])
{
	int ret = 0;

	COMMON_RESET_BEFORE_PARSE_ARGS(ulpatch_args_reset);

	ret = parse_config(argc, argv);
#if !defined(ULP_CMD_MAIN)
	if (ret == CMD_RETURN_SUCCESS_VALUE)
		return 0;
#endif
	if (ret)
		return ret;

	COMMON_IN_MAIN_AFTER_PARSE_ARGS();

	ulpatch_init();

	target_task = open_task(target_pid, FTO_ALL);

	if (!target_task) {
		fprintf(stderr, "open %d failed. %m\n", target_pid);
		return -ENOENT;
	}

	switch (command_type) {
	case CMD_PATCH:
		ret = command_patch();
		break;
	case CMD_UNPATCH:
		ret = command_unpatch();
		break;
	case CMD_NONE:
	default:
		fprintf(stderr, "What to do.\n");
	}

	close_task(target_task);
	if (patch_file)
		free(patch_file);

	return ret;
}

#if defined(ULP_CMD_MAIN)
int main(int argc, char *argv[])
{
	return ulpatch(argc, argv);
}
#endif
